flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogStateBackend.java

```diff
@@ -271,7 +273,7 @@ public StateBackend configure(ReadableConfig config, ClassLoader classLoader)
                                 keyedStateHandle instanceof ChangelogStateBackendHandle
                                         ? (ChangelogStateBackendHandle) keyedStateHandle
                                         : new ChangelogStateBackendHandleImpl(
-                                                singletonList(keyedStateHandle),
+                                                singletonMap(UUID.randomUUID(), keyedStateHandle),
```

Comment:
I see two potential problems here:
1. When up-scaling, the same handle (file) may get multipiple keys; and therefore can be discarded at some point while still in use
1. The original checkpoint does not register it's private state at all (`CompletedCheckpoint.registerSharedStatesAfterRestored`); so the original problem remains for the migration case

Am I missing something?

The only solution I see is to move this logic (`new ChangelogStateBackendHandleImpl`) to the JM.
However, this doesn't sound right because the logic is backend-specific.

Reply:
Maybe we should reconsider https://github.com/apache/flink/pull/18391#discussion_r787631347
> Btw, won't it solve the problem if we register all files in IncrementalRemoteKeyedStateHandle with SharedStateRegistry (including metadata etc.)? This would be much simpler I think in all respects.
> I guess previously there just was no need to register them (and they must be kept separate for restore).

Another option is to disallow recovery from incremental non-changelog checkpoints (leaving savepoints and full RocksDB checkpoints). But this seems quite a big limitation.

Reply:
As discussed offline, the two cases which switch jobs from not enable change-log to enabled:
> When up-scaling, the same handle (file) may get multipiple keys; and therefore can be discarded at some point while still in use

If we could introduce `UUID getIdentifier()`  to `KeyedStateHandle`, that could resolve our problems.

> The original checkpoint does not register it's private state at all (CompletedCheckpoint.registerSharedStatesAfterRestored); so the original problem remains for the migration case

I think this is only true for `CLAIM` mode. `LEGACY`  and `NO_CLAIM`  mode would set the restored checkpoint property `discardOnSubsumed` as `false`, which means it will not discard the restored checkpoint, thus the private state would not be discarded then.
Since `CLAIM` mode is not default mode, this problem is not so serious. And I create the ticket FLINK-25872 to track this problem.

Reply:
The default is `NO_CLAIM` but it's not supported (artefact copying not implemented).
`CLAIM` has the above problem.
Are you proposing to support only the `LEGACY` mode for now?

Reply:
I think you actually mean changelog state-backend does not support `NO_CLAIM` mode yet? I don't know whether we would enable changelog state-backend with `NO_CLAIM` mode in Flink-1.15.

Reply:
> I think you actually mean changelog state-backend does not support NO_CLAIM mode yet? 
Yes, that's what I mean (FLINK-25322).
Currently, it is the default (in master).

And in addition `CLAIM` mode also can not be used because of this issue in some scenarios (FLINK-25872).